micropip/_commands/freeze.py

```diff
@@ -1,12 +1,110 @@
 import importlib.metadata
 import json
+import sys
 from copy import deepcopy
-from typing import Any
+from importlib.metadata import Distribution
+from typing import TypedDict
 
 from packaging.utils import canonicalize_name
+from packaging.version import Version
 
 from .._compat import REPODATA_INFO, REPODATA_PACKAGES
 from .._utils import fix_package_dependencies
+from ..package_index import query_package
+
+IN_VENV = sys.prefix != sys.base_prefix
+
+
+class PkgEntry(TypedDict):
+    name: str
+    version: str
+    file_name: str
+    install_dir: str
+    sha256: str | None
+    imports: list[str]
+    depends: list[str]
+
+
+def get_pkg_entry_micropip(dist: Distribution, url: str) -> PkgEntry:
+    name = dist.name
+    version = dist.version
+    sha256 = dist.read_text("PYODIDE_SHA256")
+    assert sha256
+    imports = (dist.read_text("top_level.txt") or "").split()
+    requires = dist.read_text("PYODIDE_REQUIRES")
+    if not requires:
+        fix_package_dependencies(name)
+        requires = dist.read_text("PYODIDE_REQUIRES")
+    if requires:
+        depends = json.loads(requires)
+    else:
+        depends = []
+
+    return dict(
+        name=name,
+        version=version,
+        file_name=url,
+        install_dir="site",
+        sha256=sha256,
+        imports=imports,
+        depends=depends,
+    )
+
+
+async def get_pkg_entry_pip(dist: Distribution) -> PkgEntry | None:
+    resp = await query_package(dist.name)
+    ver = resp.releases.get(Version(dist.version), None)
+    if ver is None:
+        return None
+    wheel = next(ver)
+    await wheel.download({})
```

Comment:
I see this method doesn't yet exist. This call could be very expensive and would need to manage the lifecycle of the downloaded content. It may also need to re-implement things that pip does like caching or re-using pip's cache.

Reply:
Pip does expose a `list` command that will list all matching packages in the cache:

```
 ~ @ pip cache list foo.bar-0.2 --format abspath
/Users/jaraco/Library/Caches/pip/wheels/0b/ec/8b/f669ae78d94a691e6c63872909196857158b3c20f045058e7b/foo.bar-0.2-py3-none-any.whl
/Users/jaraco/Library/Caches/pip/wheels/10/f7/18/a4a94460da8bf74f841d7df4f10984b730907bea49783615a4/foo.bar-0.2-py3-none-any.whl
/Users/jaraco/Library/Caches/pip/wheels/11/9f/15/3a750be9586c80770d8355ac4e9dc3c58237b1d994f36a2a21/foo.bar-0.2-py3-none-any.whl
/Users/jaraco/Library/Caches/pip/wheels/1b/9d/16/254558b9de54233baf2ef9a35eac16ad3c23a8c5fe4122374f/foo.bar-0.2-py3-none-any.whl
/Users/jaraco/Library/Caches/pip/wheels/20/96/a7/21d787cb6f024ae978823ce36306736a68ffc7c728077edf4d/foo.bar-0.2-py3-none-any.whl
...
```

It does appear as if that "wheel cache" is only for locally-built wheels.

I haven't found a way to query for pip's http cache. Here's what pip gives for `cache info`:

```
 draft @ pip cache info
Package index page cache location (pip v23.3+): /Users/jaraco/Library/Caches/pip/http-v2
Package index page cache location (older pips): /Users/jaraco/Library/Caches/pip/http
Package index page cache size: 883.2 MB
Number of HTTP files: 10736
Locally built wheels location: /Users/jaraco/Library/Caches/pip/wheels
Locally built wheels size: 21.2 MB
Number of locally built wheels: 440
```

It's possible the `http-v2` cache is manifest using some library. If not, there's seemingly no API to access it. My guess, based on the fact that there are two cache locations, is that the http-v2 cache is reasonably stable.